demo-simple/src/main/java/com/novoda/downloadmanager/demo/MainActivity.java

```diff
@@ -60,27 +64,20 @@ public void onClick(View view) {
             }
         });
 
+        databaseMigrationUpdates = findViewById(R.id.database_migration_updates);
         View buttonMigrate = findViewById(R.id.button_migrate);
         buttonMigrate.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View view) {
-                new Thread(new Runnable() {
-                    @Override
-                    public void run() {
-                        Migrator.Callback migrationCompletionCallback = new Migrator.Callback() {
-                            @Override
-                            public void onMigrationComplete() {
-                                deleteDatabase("downloads.db");
-                            }
-                        };
-                        Migrator migrator = MigrationFactory.createVersionOneToVersionTwoMigrator(
-                                getApplicationContext(),
-                                getDatabasePath("downloads.db"),
-                                migrationCompletionCallback
-                        );
-                        migrator.migrate();
-                    }
-                }).start();
+                migrationServiceBinder.bind();
+            }
+        });
+
+        View buttonAbortMigration = findViewById(R.id.button_abort_migration);
+        buttonAbortMigration.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                migrationServiceBinder.unbind();
```

Comment:
Does this actually work? We tried it on a Pixel C and the migration still completed before the service was unbound! Had to power the device off in the end

Reply:
`unbinding` doesn't mean killing the service unfortunately. See [here](https://stackoverflow.com/questions/17146822/when-is-a-started-and-bound-service-destroyed). 

I thought that this should work when calling `stopSelf` as the above suggests but it didn't seem to work. I think that perhaps there is a leak and that is why it doesn't work. 🤔  Perhaps we need to do `stopSelf` and terminate the executor. That's assuming we want the process to stop when `unbinding`.

Reply:
I think this needs more investigation. I haven't worked on `Service`s for the longest time and I forget the best way to do them.